src/main/java/com/puppycrawl/tools/checkstyle/StatelessModule.java

```diff
@@ -0,0 +1,45 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2017 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * This annotation means that the module does not contain mutable state and
+ * may be safely used from separate threads at the same time.
+ * A module does not contain mutable state, if it does not change any field values at the execution
+ * stage (but it still can set properties at the initialization stage). The best practice is to use
+ * final fields where possible and carefully review all mutable fields. Note, that a final list
```

Comment:
do we need this best practice to be mentioned? non of Checks is following it and will never do.

"all mutable fields" .... all Checks properties are fields.

Reply:
> non of Checks is following it and will never do.

Why? All Checkstyle checks uses final fields where possible. Or do you mean 3rd party checks?

> all Checks properties are fields.

Yes, but a check may hold non-property mutable field. This note covers all fields, not only properties

Reply:
please show me example of Check with all final fields, I want to be surprised.
First check from list - https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java , almost all fields are potentially mutable.

Reply:
@romani This is not exactly what I mean. A "field" does not necessary mean "property". E.g. [AbstractSuperCheck](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/AbstractSuperCheck.java) contains a final field, which holds a check context. 

Here is a check with final property field: [IllegalCatchCheck](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheck.java).

Both properties are mutable, but they are marked as `final`. This is what I mean by "All Checkstyle checks uses final fields where possible. "

I could remove the note about "best practice" from the javadoc if needed

Reply:
please remove that note at all, this javadoc is not post of best practices in MT

Reply:
Fixed